### ISSUE TITLE
No nonnegativity constraints are imposed with all_outcomes=True and feas_as_optim=True

### ISSUE THREAD

**Issue by eliewolfe** [ecboghiu/inflation#145]
When running LPs with include_all_outcomes=True, both bisection and dual-certificate subalgorithms fail when doing a max_within_feasible optimization.
The "dual" method false returns visibility=1, and the bisection method throws an error about not changing sign across the vis=0 and vis=1 iterations.

```
import numpy as np
from inflation import InflationProblem, InflationLP, max_within_feasible
from sympy import Symbol
vis = Symbol("v")

p_AB_giv_XY=np.zeros((3,3,3,3))

p_AB_giv_XY[0,0,0,0] = 1/4
p_AB_giv_XY[0,2,0,0] = 1/4
p_AB_giv_XY[2,0,0,0] = 1/4
p_AB_giv_XY[1,1,0,0] = 1/4

for (x,y) in {(0,1), (0,2), (1,0), (2,0)}:
    p_AB_giv_XY[0,0,x,y] = 1/2
    p_AB_giv_XY[2,2,x,y] = 1/4
    p_AB_giv_XY[1,1,x,y] = 1/4

for y in {1,2}:
    p_AB_giv_XY[0,0,1,y] = 1/4
    p_AB_giv_XY[0,1,1,y] = 1/4
    p_AB_giv_XY[1,0,1,y] = 1/4
    p_AB_giv_XY[2,2,1,y] = 1/4

p_AB_giv_XY[0,0,2,1] = 1/4
p_AB_giv_XY[2,0,2,1] = 1/4
p_AB_giv_XY[0,1,2,1] = 1/4
p_AB_giv_XY[1,2,2,1] = 1/4

p_AB_giv_XY[0,1,2,2] = 1/4
p_AB_giv_XY[0,2,2,2] = 1/4
p_AB_giv_XY[1,0,2,2] = 1/4
p_AB_giv_XY[2,0,2,2] = 1/4

p_det = np.zeros((3,3,3,3))
p_det[0,0,:,:] = 1

p_local = np.zeros((3,3,3,3))
p_local[0,0,:,:] = 1/2
p_local[0,1,:,:] = 1/8
p_local[0,2,:,:] = 1/8
p_local[1,0,:,:] = 1/8
p_local[2,0,:,:] = 1/8

LocalityCheck = InflationProblem(dag={"rho_AB": ["A", "B"]},
                                    classical_sources=["rho_AB"],
                                    outcomes_per_party=(3,3),
                                    settings_per_party=(3,3),
                                    inflation_level_per_source=(1,),
                                    order=["A", "B"])
BellTest = InflationLP(LocalityCheck,
                    verbose=0,
                    include_all_outcomes=False)
BellTest.set_distribution(vis*p_AB_giv_XY+(1-vis)*p_local)
v_max = max_within_feasible(BellTest, BellTest.known_moments, "dual")
print("local visibility (include all outcomes):", v_max)
BellTest = InflationLP(LocalityCheck,
                    verbose=0,
                    include_all_outcomes=True)
BellTest.set_distribution(vis*p_AB_giv_XY+(1-vis)*p_local)
v_max = max_within_feasible(BellTest, BellTest.known_moments, "dual")
print("local visibility (Collins-Gisin notation):", v_max)
```

**Comment by apozas** [comment]
For the distribution that you are inputting, I am getting very different values for `vis=1`. For `include_all_outcomes=False` I am getting a maximum lambda (i.e., the objective value when solving with `feas_as_optim=True`) of `-0.00111607142857333`, while for `include_all_outcomes=True` I get `1.5540715151804617e-16`. So this is not a problem of `max_within_feasible`. By any chance, do we know if the distribution is local or not, in order to know which is the failing program?

**Comment by apozas** [comment]
The `maximize_via_xxx` functions call, for numerical stability, the `solve` function with `feas_as_optim=True`. When `feas_as_optim` is set to `True`, `InflationLP.solve` [overrides `relax_inequalities`](https://github.com/ecboghiu/inflation/blob/a74e88ec7d3e0b437e3882854021b6af0c4c0bce/inflation/lp/InflationLP.py#L599) and sets it to `True`. And this value overrides `default_non_negative` in [`solveLP_sparse`](https://github.com/ecboghiu/inflation/blob/a74e88ec7d3e0b437e3882854021b6af0c4c0bce/inflation/lp/lp_utils.py#L208).

The result is that the constraints that all probabilities must be positive are not imposed. Indeed, when solving the feasibility problem with `feas_as_optim=True` and, for instance, `vis=1`, one sees negative probabilities in the solution.

When `include_all_outcomes=False`, the relations between the probabilities are encoded in `moment_inequalities`, and when `include_all_outcomes=True`, these relations are stored in `moment_equalities`.
It seems like these relations between probabilities in inequality form (when using `include_all_outcomes=False`) contain some positivity constraints. By inspection of the object `inequalities` that is passed to `solveLP_sparse`, I can see that there are some rows which only contain one 1. So some probabilities are being imposed to be non-negative, and this seems enough to be enough to constrain the problem.

I believe that the solution to this would be not to set `default_non_negative` to `False` whenever we have `feas_as_optim`, (which results in not imposing any non-negativity constraints at all), and instead relax the non-negativity constraints, as [we do with the rest of inequality constraints](https://github.com/ecboghiu/inflation/blob/a74e88ec7d3e0b437e3882854021b6af0c4c0bce/inflation/lp/lp_utils.py#L249). Does this look like a good course of action?

**Comment by eliewolfe** [comment]
I've done a bunch of debugging, and I agree that the issue stems from setting `default_non_negative` to `False` when `feas_as_optim=True`. However, that is not a good fix.
The problem arises because the current code _fails to generate inequalities_ whenever `all_outcomes=True`. 
If we generated the inequalities correctly, then we could indeed just relax the inequality constraints and all would be well. I'm looking into adjusting the inequality generation in `InflationLP`.

**Comment by eliewolfe** [comment]
Fixed in commit [b0f1497](https://github.com/ecboghiu/inflation/commit/b0f149766b2d6c8bd8944df51b9a94c847cec628).
